File: TUnit.TestProject/UniqueBuilderContextsOnEnumerableDataGeneratorTests.cs
using Shouldly;
using TUnit.TestProject.Attributes;

namespace TUnit.TestProject;

[EngineTest(ExpectedResult.Pass)]
public class UniqueBuilderContextsOnEnumerableDataGeneratorTests
{
    [Test, UniqueBuilderContextsOnEnumerableDataGeneratorTestsGenerator]
    public void Test(int value)
    {
    }
}

public class UniqueBuilderContextsOnEnumerableDataGeneratorTestsGenerator : DataSourceGeneratorAttribute<int>
{
    protected override IEnumerable<Func<int>> GenerateDataSources(DataGeneratorMetadata dataGeneratorMetadata)
    {
        var id1 = dataGeneratorMetadata.TestBuilderContext.Current.Id;
        var id2 = dataGeneratorMetadata.TestBuilderContext.Current.Id;

        yield return () => 1;

        var id3 = dataGeneratorMetadata.TestBuilderContext.Current.Id;

        yield return () => 2;

        var id4 = dataGeneratorMetadata.TestBuilderContext.Current.Id;

        id1.ShouldBe(id2);

        id3.ShouldNotBe(id1);

        id4.ShouldNotBe(id1);
        id4.ShouldNotBe(id3);
    }
}

File: TUnit.TestProject.FSharp/UniqueBuilderContextsOnEnumerableDataGeneratorTests.fs
namespace TUnit.TestProject.FSharp

open System.Collections.Generic
open TUnit.Core

// Equivalent of UniqueBuilderContextsOnEnumerableDataGeneratorTests.cs

type UniqueBuilderContextsOnEnumerableDataGeneratorTests() =
    [<Test; UniqueBuilderContextsOnEnumerableDataGeneratorTestsGenerator>]
    member _.Test(value: int) = ()

and UniqueBuilderContextsOnEnumerableDataGeneratorTestsGenerator() =
    inherit DataSourceGeneratorAttribute<int>()
    override _.GenerateDataSources(dataGeneratorMetadata) =
        seq {
            let id1 = dataGeneratorMetadata.TestBuilderContext.Current.Id
            let id2 = dataGeneratorMetadata.TestBuilderContext.Current.Id
            yield (fun () -> 1)
            let id3 = dataGeneratorMetadata.TestBuilderContext.Current.Id
            yield (fun () -> 2)
            let id4 = dataGeneratorMetadata.TestBuilderContext.Current.Id
            // ShouldBe/ShouldNotBe logic can be implemented with F# assertions or left as comments
        }
